controllers/drpolicy_controller.go

```diff
@@ -213,221 +200,76 @@ func hasConflictingDRPolicy(match *ramen.DRPolicy, list ramen.DRPolicyList) erro
-		if haveOverlappingMetroZones(match, drp) {
+		if haveOverlappingMetroZones(match, drp, drclusters) {
 			return fmt.Errorf("drpolicy: %v has overlapping metro region with another drpolicy %v", match.Name, drp.Name)
 		}
 	}
 
 	return nil
 }
 
-func validateManagedClusters(ctx context.Context, apiReader client.Reader, drpolicy *ramen.DRPolicy) error {
+func validateManagedClusters(ctx context.Context,
+	apiReader client.Reader,
+	drpolicy *ramen.DRPolicy,
+	drclusters *ramen.DRClusterList) error {
 	drpolicies, err := util.GetAllDRPolicies(ctx, apiReader)
 	if err != nil {
 		return fmt.Errorf("validate managed cluster in drpolicy %v failed: %w", drpolicy.Name, err)
 	}
 
-	err = hasConflictingDRPolicy(drpolicy, drpolicies)
+	err = hasConflictingDRPolicy(drpolicy, drclusters, drpolicies)
 	if err != nil {
 		return fmt.Errorf("validate managed cluster in drpolicy failed: %w", err)
 	}
 
 	return nil
 }
 
-func validateS3Profiles(ctx context.Context, apiReader client.Reader,
-	objectStoreGetter ObjectStoreGetter, drpolicy *ramen.DRPolicy, listKeyPrefix string, log logr.Logger) (string, error) {
-	for i := range drpolicy.Spec.DRClusterSet {
-		cluster := &drpolicy.Spec.DRClusterSet[i]
-		if cluster.ClusterFence == ramen.ClusterFenceStateFenced ||
-			cluster.ClusterFence == ramen.ClusterFenceStateManuallyFenced {
-			continue
-		}
-
-		if reason, err := s3ProfileValidate(ctx, apiReader, objectStoreGetter,
-			cluster.S3ProfileName, listKeyPrefix, log); err != nil {
-			return reason, err
-		}
-	}
-
-	return "", nil
-}
-
-func s3ProfileValidate(ctx context.Context, apiReader client.Reader,
-	objectStoreGetter ObjectStoreGetter, s3ProfileName, listKeyPrefix string,
-	log logr.Logger,
-) (string, error) {
-	objectStore, err := objectStoreGetter.ObjectStore(ctx, apiReader, s3ProfileName, "drpolicy validation", log)
-	if err != nil {
-		return "s3ConnectionFailed", fmt.Errorf("%s: %w", s3ProfileName, err)
-	}
-
-	if _, err := objectStore.ListKeys(listKeyPrefix); err != nil {
-		return "s3ListFailed", fmt.Errorf("%s: %w", s3ProfileName, err)
-	}
-
-	return "", nil
-}
-
-func validateCIDRsFormat(drpolicy *ramen.DRPolicy, log logr.Logger) error {
-	// validate the CIDRs format
-	for i := range drpolicy.Spec.DRClusterSet {
-		cluster := &drpolicy.Spec.DRClusterSet[i]
-		if err := ParseCIDRs(cluster, log); err != nil {
-			return err
-		}
-	}
-
-	return nil
-}
-
-func ParseCIDRs(cluster *ramen.ManagedCluster, log logr.Logger) error {
-	invalidCidrs := []string{}
-
-	for i := range cluster.CIDRs {
-		if _, _, err := net.ParseCIDR(cluster.CIDRs[i]); err != nil {
-			invalidCidrs = append(invalidCidrs, cluster.CIDRs[i])
-
-			log.Error(err, ReasonValidationFailed)
-		}
-	}
-
-	if len(invalidCidrs) > 0 {
-		return fmt.Errorf("cluster %s has invalid CIDRs %s", cluster.Name, strings.Join(invalidCidrs, ", "))
-	}
-
-	return nil
-}
-
-type objectUpdater struct {
+type drpolicyUpdater struct {
```

Comment:
NOTE: The type is made Kind specific as accessing <kind>.Status is not generic and depends on the type. It is conceivable that we can write functions in the base type and create a new interface and hence retain a single type here, but the interfaces would get exposed in the same package as v1alpha1 for ramen, hence did not do that.

Changing it so and adding some generic routines seemed the optimal middle path for this refactor.